streaming/storage/deque.go

```diff
@@ -0,0 +1,383 @@
+	*/
+	ll.firstElement = 0
+	ll.lastElement = 1
+	ll.initialized = false
+
+	return nil
+}
+
+func (ll *Deque) GetIterator() *DequeIterator {
+	it := ll.tx.WithPrefix(dequeValueKeyPrefix).Iterator(WithDefault())
+	return NewDequeIterator(it)
+}
+
+func (lli *DequeIterator) Next(value proto.Message) error {
+	err := lli.it.Next(value)
+	if err == ErrEndOfIterator || err == nil {
+		return err
+	}
+
+	return errors.Wrap(err, "couldn't read next element")
+}
+
+func (lli *DequeIterator) Close() error {
+	return lli.it.Close()
+}
+
+func getIndexKey(index int) []byte {
+	bytes := make([]byte, 0)
+	bytes = append(bytes, dequeValueKeyPrefix...)
+	bytes = append(bytes, octosql.MonotonicMarshalInt64(int64(index))...)
+
+	return bytes
+}
+
+func (ll *Deque) initialize() error {
+	last, err := ll.getLast()
+	if err != nil {
+		return errors.Wrap(err, "couldn't initialize the queue last attribute")
+	}
+
+	first, err := ll.getFirst()
+	if err != nil {
+		return errors.Wrap(err, "couldn't initialize the queue first attribute")
+	}
+
+	ll.lastElement = last
+	ll.firstElement = first
+	ll.initialized = true
+
+	return nil
+}
+
+func (ll *Deque) getLast() (int, error) {
+	return ll.getAttribute(dequeLastElementKey)
+}
+
+func (ll *Deque) getFirst() (int, error) {
+	return ll.getAttribute(dequeFirstElementKey)
+}
+
+func (ll *Deque) getAttribute(attr []byte) (int, error) {
+	value, err := ll.tx.Get(attr)
+	stringAttribute := string(attr)
+	switch err {
+	case badger.ErrKeyNotFound:
+		value, ok := defaultValues[stringAttribute]
+		if !ok {
+			return 0, errors.New("this byte slice isn't a queue attribute")
+		}
+
+		err2 := ll.tx.Set(attr, octosql.MonotonicMarshalInt64(int64(value)))
+		if err2 != nil {
+			return 0, errors.Wrapf(err2, "couldn't initialize queue %s field", stringAttribute)
+		}
+
+		return value, nil
+	case nil:
+		i, err := octosql.MonotonicUnmarshalInt64(value)
+		return int(i), err
+	default:
+		return 0, errors.Wrapf(err, "couldn't read %s of queue", stringAttribute)
+	}
+}
+
+func (ll *Deque) incLast() error {
+	err := ll.incAttribute(dequeLastElementKey)
+	if err != nil {
+		return errors.Wrap(err, "failed to increase queue last attribute")
+	}
+
+	ll.lastElement += 1
+	return nil
+}
+
+func (ll *Deque) incFirst() error {
+	err := ll.incAttribute(dequeFirstElementKey)
+	if err != nil {
+		return errors.Wrap(err, "failed to increase queue first attribute")
+	}
+
+	ll.firstElement += 1
+	return nil
+}
+
+func (ll *Deque) decLast() error {
+	err := ll.decAttribute(dequeLastElementKey)
+	if err != nil {
+		return errors.Wrap(err, "failed to decrease queue last attribute")
+	}
+
+	ll.lastElement -= 1
+	return nil
+}
+
+func (ll *Deque) decFirst() error {
+	err := ll.decAttribute(dequeFirstElementKey)
+	if err != nil {
+		return errors.Wrap(err, "failed to decrease queue first attribute")
+	}
+
+	ll.firstElement -= 1
+	return nil
+}
+
+func (ll *Deque) incAttribute(attr []byte) error {
```

Comment:
I think you can move this function between incFirst and decLast so that it improve readability